### ISSUE TITLE
vinyl: vy_apply_upsert does not satisfy associativity rule

### ISSUE THREAD

**Issue by alyapunov** [tarantool/tarantool#5105]
Now result of several upserts can depend on the order of their application.
<details>
<summary>correct behaviour</summary>

```lua
tarantool> s = box.schema.space.create('test', {engine='vinyl'})
---
...

tarantool> i = s:create_index('pk', {parts={2, 'uint'}}) -- note that the index is by second field
---
...

tarantool> s:replace{1, 2, 3, 'default'}
---
- [1, 2, 3, 'default']
...

tarantool> s:upsert({2, 2, 2}, {{'=', 4, 'upserted'}})
---
...

tarantool> s:select{}
---
- - [1, 2, 3, 'upserted'] -- good
...

tarantool> s:upsert({2, 2, 2}, {{'#', 1, 1}, {'!', 3, 1}}) -- this upsert will fail and thus ignored
---
...

tarantool> s:select{}
---
- - [1, 2, 3, 'upserted'] -- nothing changed as expected
...
```
</details>

<details>
<summary>incorrect behaviour</summary>

```lua
tarantool> s = box.schema.space.create('test', {engine='vinyl'})
---
...

tarantool> i = s:create_index('pk', {parts={2, 'uint'}}) -- note that the index is by second field
---
...

tarantool> s:replace{1, 2, 3, 'default'}
---
- [1, 2, 3, 'default']
...

tarantool> box.snapshot()
---
- ok
...

tarantool> s:upsert({2, 2, 2}, {{'=', 4, 'upserted'}})
---
...

tarantool> s:upsert({2, 2, 2}, {{'#', 1, 1}, {'!', 3, 1}}) -- this upsert will fail and thus ignored
---
...

tarantool> box.snapshot()
---
- ok
...

s:select{}
---
- - [1, 2, 3, 'default'] -- oops, the first upsert was ignored too
...
```
</details>

Long time ago several when upserts was joined they preserved 'series of ops` of each parent upsert, and that worked correctly. Broken by f4615e732a28ea8dfb9b1ebecb5e539dc5cd48d3 and #1834

**Comment by Korablev77** [comment]
A-ha, I've also noticed this misbehavior but considered it to be a feature. The thing are going to be even worse in scope of https://github.com/tarantool/tarantool/commit/7c98fabb09bef52546b592ce5603e7e69b3ffca1 :

> The only drawback of this approach is that if one of upserts *really*
> modifies PK of tuple stored on disk, all squashed operations will be
> skipped as well. For example:
> 
> s:upsert({1}, {{'+', 2, 100}})
> s:upsert({1}, {{'=', -2, 200}})
>

**Comment by alyapunov** [comment]
Yeah, I found it during #1622 review.

**Comment by alyapunov** [comment]
I think we could fix it along with #5092
